File: src/test/java/eotree/TestEOObject.java
package eotree;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static util.ListUtils.listOf;

import arrow.core.Some;
import org.junit.jupiter.api.Test;

/**
 * EO object tests.
 */
public class TestEOObject {

    private static final String memAtom = "memory";

    @Test
    public void testGenerateEOZeroIndent() {
        var object = new EOObject(
                listOf(
                        "free1",
                        "free2"
                ),
                new Some<>("vararg"),
                listOf(
                        new EOBndExpr(
                                new EODot(memAtom),
                                "bnd1"
                        ),
                        new EOBndExpr(
                                new EODot(memAtom),
                                "bnd2"
                        )
                ),
                "comment"
        );
        assertEquals(
                "# comment\n" +
                "[free1 free2 vararg...]\n" +
                "  memory > bnd1\n" +
                "  memory > bnd2",
                object.generateEO(0)
        );
    }

    @Test
    public void testGenerateEONonZeroIndent() {
        var object = new EOObject(
                listOf("free1", "free2"),
                new Some<>("vararg"),
                listOf(
                        new EOBndExpr(
                                new EODot(memAtom),
                                "bnd1"
                        ),
                        new EOBndExpr(
                                new EODot(memAtom),
                                "bnd2"
                        )
                ),
                "comment"
        );
        //noinspection TextBlockMigration
        assertEquals(
                "  # comment\n"
                + "  [free1 free2 vararg...]\n"
                + "    memory > bnd1\n"
                + "    memory > bnd2",
                object.generateEO(1)
        );
    }
}

File: src/test/kotlin/eotree/TestEOObject.kt
package eotree

import arrow.core.Some
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.polystat.j2eo.eotree.EOBndExpr
import org.polystat.j2eo.eotree.EODot
import org.polystat.j2eo.eotree.EOObject
import org.polystat.j2eo.util.ListUtils.listOf

/**
 * EO object tests.
 */
class TestEOObject {
    @Test
    fun testGenerateEOZeroIndent() {
        val eoObject = EOObject(
            listOf(
                "free1",
                "free2"
            ),
            Some("vararg"),
            listOf(
                EOBndExpr(
                    EODot(memAtom),
                    "bnd1"
                ),
                EOBndExpr(
                    EODot(memAtom),
                    "bnd2"
                )
            ),
            "comment"
        )
        Assertions.assertEquals(
            "# comment\n" +
                "[free1 free2 vararg...]\n" +
                "  memory > bnd1\n" +
                "  memory > bnd2",
            eoObject.generateEO(0)
        )
    }

    @Test
    fun testGenerateEONonZeroIndent() {
        val eoObject = EOObject(
            listOf("free1", "free2"),
            Some("vararg"),
            listOf(
                EOBndExpr(
                    EODot(memAtom),
                    "bnd1"
                ),
                EOBndExpr(
                    EODot(memAtom),
                    "bnd2"
                )
            ),
            "comment"
        )
        Assertions.assertEquals(
            """  # comment
  [free1 free2 vararg...]
    memory > bnd1
    memory > bnd2""",
            eoObject.generateEO(1)
        )
    }

    companion object {
        private const val memAtom = "memory"
    }
}
